File: cf/75e.c
#include <math.h>
#include <stdio.h>

typedef struct { int x, y; } point;

void read(point *p) {
	scanf("%d%d", &p->x, &p->y);
}

typedef struct { int x, y; } vector;

int cross(vector v, vector w) {
	return v.x*w.y - w.x*v.y;
}

int dot(vector v, vector w) {
	return v.x*w.x + v.y*w.y;
}

vector sub(point a, point b) {
	return (vector){b.x-a.x, b.y-a.y};
}

typedef struct { point a, b; } segment;

int parallels(segment s, segment t) {
	return cross(sub(s.a, s.b), sub(t.a, t.b)) == 0;
}

int direction(point a, point b, point c) {
	return cross(sub(a, c), sub(a, b));
}

int min(int a, int b) { return a < b ? a : b; }
int max(int a, int b) { return a > b ? a : b; }

int on_segment(point a, point b, point c) {
	return min(a.x, b.x) <= c.x && c.x <= max(a.x, b.x) &&
		min(a.y, b.y) <= c.y && c.y <= max(a.y, b.y);
}

int intersects(segment s, segment t) {
	int d1 = direction(t.a, t.b, s.a),
	    d2 = direction(t.a, t.b, s.b),
	    d3 = direction(s.a, s.b, t.a),
	    d4 = direction(s.a, s.b, t.b);

	return (d1 > 0 && d2 < 0 || d1 < 0 && d2 > 0) &&
			(d3 > 0 && d4 < 0 || d3 < 0 && d4 > 0) ||
		d1 == 0 && on_segment(t.a, t.b, s.a) ||
		d2 == 0 && on_segment(t.a, t.b, s.b) ||
		d3 == 0 && on_segment(s.a, s.b, t.a) ||
		d4 == 0 && on_segment(s.a, s.b, t.b);
}

typedef struct { double x, y; } float_point;

float_point intersection(segment s, segment t) {
	double c = 1. * cross(sub(s.a, t.a), sub(t.a, t.b)) / cross(sub(s.a, s.b), sub(t.a, t.b));
	return (float_point){s.a.x + c*(s.b.x - s.a.x), s.a.y + c*(s.b.y - s.a.y)};
}

double sqr(double a) { return a*a; }

float_point to_float_point(point p) {
	return (float_point){p.x, p.y};
}

double dist(float_point a, float_point b) {
	return sqrt(sqr(b.x-a.x) + sqr(b.y-a.y));
}

int main() {
	point p1, p2; read(&p1); read(&p2);

	int n; scanf("%d", &n);
	point ps[n];
	for (int i = 0; i < n; i++)
		read(&ps[i]);

	segment s = {p1, p2};
	int c = 0, i1, i2;
	float_point ip1, ip2;
	for (int i = 0; i < n; i++) {
		segment t = {ps[i], ps[(i+1)%n]};

		if (!intersects(s, t) || cross(sub(p1, p2), sub(p1, t.a)) == 0)
			continue;

		if (parallels(s, t)) {
			c++;
			continue;
		}

		float_point p = intersection(s, t);
		if (c == 0) {
			i1 = i;
			ip1 = p;
		} else if (c == 1) {
			i2 = i;
			ip2 = p;
		}
		c++;
	}


	float_point fp1 = to_float_point(p1), fp2 = to_float_point(p2);

	if (c != 2) {
		printf("%lf\n", dist(fp1, fp2));
		return 0;
	}

	if (dist(fp1, ip2) < dist(fp1, ip1)) {
		float_point tp = ip1; ip1 = ip2; ip2 = tp;
		int t = i1; i1 = i2; i2 = t;
	}

	double m = 2*dist(ip1, ip2);

	float_point fps[n];
	for (int i = 0; i < n; i++)
		fps[i] = to_float_point(ps[i]);

	double d = dist(ip1, fps[(i1+1)%n]);
	for (int i = (i1+1)%n; i != i2; i = (i+1)%n) {
		d += dist(fps[i], fps[(i+1)%n]);
	}
	d += dist(fps[i2], ip2);
	if (d < m)
		m = d;

	d = dist(ip1, fps[i1]);
	for (int i = i1; (i-1+n)%n != i2; i = (i-1+n)%n)
		d += dist(fps[i], fps[(i-1+n)%n]);
	d += dist(fps[(i2+1)%n], ip2);
	if (d < m)
		m = d;

	printf("%lf\n", m + dist(fp1, ip1) + dist(ip2, fp2));

	return 0;
}

File: cf/75e.hs
import Control.Applicative
import Data.Function
import Data.List

data Point = Point Double Double deriving (Eq)
data Segment = Segment { segmentStart, segmentEnd :: Point } deriving (Eq)
data Vector = Vector Double Double

getWords = map read . words <$> getLine

splitEvery n = unfoldr $ \xs -> if null xs then Nothing else Just $ splitAt n xs
getPoints = map (\[a, b] -> Point a b) . splitEvery 2 <$> getWords

cross (Vector x1 y1) (Vector x2 y2) = x1*y2 - x2*y1
dot (Vector x1 y1) (Vector x2 y2) = x1*x2 + y1*y2
sub (Point px py) (Point qx qy) = Vector (qx - px) (qy - py)

intersects s@(Segment p1 p2) t@(Segment p3 p4) =
		(d1 > 0 && d2 < 0 || d1 < 0 && d2 > 0) &&
		(d3 > 0 && d4 < 0 || d3 < 0 && d4 > 0) ||
		d1 == 0 && onSegment p1 t ||
		d2 == 0 && onSegment p2 t ||
		d3 == 0 && onSegment p3 s ||
		d4 == 0 && onSegment p4 s
	where
		direction p q r = sub p r `cross` sub p q

		onSegment (Point rx ry) (Segment (Point px py) (Point qx qy)) =
			min px qx <= rx && rx <= max px qx &&
			min py qy <= ry && ry <= max py qy

		d1 = direction p3 p4 p1
		d2 = direction p3 p4 p2
		d3 = direction p1 p2 p3
		d4 = direction p1 p2 p4

data Line = Line Point Point

onLine r (Line p q) = sub p q `cross` sub p r == 0

intersectsRay s@(Segment p1 p2) t@(Segment p3 p4) =
	intersects s t && not (p3 `onLine` Line p1 p2)

intersection s@(Segment sa sb) t@(Segment ta tb) = 
		sa `add` (c `mult` (sa `sub` sb))
	where
		mult a (Vector x y) = Vector (a * x) (a * y)
		add (Point px py) (Vector vx vy) = Point (px + vx) (py + vy)
		c = (sub sa ta `cross` sub ta tb) / (sub sa sb `cross` sub ta tb)

distance (Point px py) (Point qx qy) = sqrt $ (qx - px) ** 2 + (qy - py) ** 2

solve start end polygon
		| length intersecting < 2 = start `distance` end
		| otherwise =
			start `distance` i1 +
			minimum [distThrough, distAround1, distAround2] +
			i2 `distance` end
	where
		startEnd = Segment start end
		segments = zipWith Segment polygon $ tail polygon ++ [head polygon]

		intersecting = filter (startEnd `intersectsRay`) segments
		intersections = map (intersection startEnd) intersecting

		(s1,i1):(s2,i2):_ = sortBy (compare `on` (distance start . snd)) $
				intersecting `zip` intersections

		distThrough = 2 * (i1 `distance` i2)

		distAround s1 s2 segments = sum $ map (\(Segment p q) -> p `distance` q) $
			takeWhile (/= s2) $ tail $ dropWhile (/= s1) $ segments ++ segments

		distAround1 = 
			i1 `distance` segmentEnd s1 +
			distAround s1 s2 segments +
			segmentStart s2 `distance` i2

		distAround2 = 
			i1 `distance` segmentStart s1 +
			distAround s1 s2 (reverse segments) +
			segmentEnd s2 `distance` i2

main = do
	[start, end] <- getPoints
	_ <- getLine
	polygon <- getPoints
	print $ solve start end polygon

